app/Platform/Services/TriggerDecoderService.php

```diff
@@ -0,0 +1,552 @@
+                case 'r': case 'R': $flag = 'Reset If'; break;
+                case 'a': case 'A': $flag = 'Add Source'; $scalable = true; break;
+                case 'b': case 'B': $flag = 'Sub Source'; $scalable = true; break;
+                case 'c': case 'C': $flag = 'Add Hits'; break;
+                case 'd': case 'D': $flag = 'Sub Hits'; break;
+                case 'n': case 'N': $flag = 'And Next'; break;
+                case 'o': case 'O': $flag = 'Or Next'; break;
+                case 'm': case 'M': $flag = 'Measured'; break;
+                case 'q': case 'Q': $flag = 'Measured If'; break;
+                case 'i': case 'I': $flag = 'Add Address'; $scalable = true; break;
+                case 't': case 'T': $flag = 'Trigger'; break;
+                case 'z': case 'Z': $flag = 'Reset Next If'; break;
+                case 'g': case 'G': $flag = 'Measured %'; break;
+                default: $flag = $mem[0]; break;
+            }
+
+            $mem = substr($mem, 2);
+        }
+
+        [$lType, $lSize, $lMemory, $mem] = $this->parseOperand($mem);
+
+        if (strlen($mem) === 0) {
+            // no operator
+        } elseif ($scalable && !$this->isScalerOperator($mem[0])) {
+            $mem = ''; // non-scaler operator ignored for scalable operations
+        } else {
+            $cmp = $mem[0];
+            $cmplen = 1;
+            switch ($mem[0]) {
+                case '=':
+                    if ($mem[1] === '=') {
+                        $cmplen = 2;
+                    }
+                    break;
+
+                case '!':
+                    if ($mem[1] === '=') {
+                        $cmp = '!=';
+                        $cmplen = 2;
+                    }
+                    break;
+
+                case '<':
+                    if ($mem[1] === '=') {
+                        $cmp = '<=';
+                        $cmplen = 2;
+                    }
+                    break;
+
+                case '>':
+                    if ($mem[1] === '=') {
+                        $cmp = '>=';
+                        $cmplen = 2;
+                    }
+                    break;
+            }
+            $mem = substr($mem, $cmplen);
+
+            [$rType, $rSize, $rMemVal, $mem] = $this->parseOperand($mem);
+
+            $hits = $scalable ? '' : '0';
+            if (strlen($mem) > 0 && ($mem[0] === '(' || $mem[0] === '.')) {
+                $hits = substr($mem, 1, strlen($mem) - 2);
+            }
+        }
+
+        return [
+            'Flag' => $flag,
+            'SourceType' => $lType,
+            'SourceSize' => $lSize,
+            'SourceAddress' => $lMemory,
+            'Operator' => $cmp,
+            'TargetType' => $rType,
+            'TargetSize' => $rSize,
+            'TargetAddress' => $rMemVal,
+            'HitTarget' => $hits,
+        ];
+    }
+
+    public function decode(string $serializedTrigger): array
```

Comment:
`->map()` should eliminate the need for writing nested `for` loops:

```php
public function decode(string $serializedTrigger): array
{
    $serializedGroups = collect(preg_split("/(?<!0x)[S$]/", $serializedTrigger));

    return $serializedGroups->map(function ($groupString, $index) {
        $reqs = explode('_', $groupString);
        $isIndirect = false;

        $conditions = collect($reqs)->filter()->map(function ($req) use (&$isIndirect) {
            $condition = $this->parseCondition($req);
            $condition['IsIndirect'] = $isIndirect;

            if ($condition['SourceType'] === 'Value') {
                $condition['SourceTooltip'] = hexdec($condition['SourceAddress']);
            }
            if ($condition['TargetType'] === 'Value') {
                $condition['TargetTooltip'] = hexdec($condition['TargetAddress']);
            }

            $isIndirect = $condition['Flag'] === 'Add Address';
            return $condition;
        });

        return [
            'Label' => ($index === 0) ? 'Core Group' : "Alt Group $index",
            'Conditions' => $conditions->all()
        ];
    })->all();
}
```

Reply:
The only thing I'm shooting for is readability and maintainability. I agree with your points and believe the inner `foreach` is a good change.